<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Subset sequences from sq objects — bite • tidysq</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Subset sequences from sq objects — bite" />
<meta property="og:description" content="Extracts a defined range of elements (amino acids or nucleotides) 
from a sequence." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidysq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/michbur/tidysq/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Subset sequences from sq objects</h1>
    <small class="dont-index">Source: <a href='https://github.com/michbur/tidysq/blob/master/R/bite.R'><code>R/bite.R</code></a></small>
    <div class="hidden name"><code>bite.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Extracts a defined range of elements (amino acids or nucleotides) 
from a sequence.</p>
    </div>

    <pre class="usage"><span class='fu'>bite</span>(<span class='kw'>sq</span>, <span class='kw'>indices</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sq</th>
      <td><p>a <code><a href='sq.html'>sq</a></code> object.</p></td>
    </tr>
    <tr>
      <th>indices</th>
      <td><p>a <code><a href='https://rdrr.io/r/base/numeric.html'>numeric</a></code> vector of subsequence indices to extract from
each sequence. The function follows the normal R conventions for indexing 
vectors, including negative indices.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code><a href='sq.html'>sq</a></code> object of the same type as input sq, where each 
element is a subsequence created by indexing corresponding sequence from 
input sq object with input indices.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Amino acids and nucleic acid sequences are represented as <code><a href='sq.html'>sq</a></code> 
object in the <code><a href='tidysq-package.html'>tidysq</a></code> package. Often one needs to get only a 
single letter, or the sequence of a defined range from the original sequences. 
A subsequence is a sequence that can be derived from the original sequence 
by trimming some elements (letters) without changing the order of the 
remaining elements. To get a subsequence from each sequence contained in 
the <code><a href='sq.html'>sq</a></code> object with the same indices. This is for example 
useful to extract a user-defined region from a sequence.</p>
<p>The usage of <code>bite</code> follows the normal R conventions. For details 
refer to the R documentation (see 
<a href='https://cran.r-project.org/doc/manuals/r-release/R-intro.html#Index-vectors'>https://cran.r-project.org/doc/manuals/r-release/R-intro.html#Index-vectors</a>). 
Extracting indices not present in the sequence results in introducing 
<code><a href='https://rdrr.io/r/base/NA.html'>NA</a></code> (‘Not Available’ / Missing Value) values. 
Information about it is printed on a console depending on the value of option 
'tidysq_a_bite_na' - it can be either a warning (default), an error,
a message or no information (you can check details in <code><a href='tidysq-options.html'>tidysq-options</a>)</code>. 
<code>NA</code> values can be removed by using <code><a href='remove_na.html'>remove_na</a></code> function.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='sq.html'>sq</a></code> <code><a href='remove_na.html'>remove_na</a></code> <code><a href='tidysq-options.html'>tidysq-options</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Creating object, called sq to work on:</span>
<span class='co'># The first four sequences of are miRNAs and isomiRs that were taken from </span>
<span class='co'># Tan et al. (2014) Nucleic Acid Research, doi: 10.1093/nar/gku656. </span>
<span class='co'># Selected miR-9 (isomiRs) sequences from neural progenitor stem cells</span>
<span class='co'># look as follows: TCTTTGGTTATCTAGCTGTATGA, CTTTGGTTATCTAGCTGTATGA, </span>
<span class='co'># TCTTTGGTTATCTAGCTGTATG, TCTTTGGTTATCTAGCTGTATGAA</span>
<span class='co'># The remaining sequences represent short artificial random sequences.</span>
<span class='co'># Tan et al. have shown that "majority of miRNA genes encode mature isomers </span>
<span class='co'># that vary in size by one or more bases at the 3' and/or 5' end of the miRNA.</span>
<span class='co'># In the following the bite function is used to trim nucleotieds 5' for a </span>
<span class='co'># simple visual alignment.</span>
<span class='co'># construct_sq is used to create a sq object</span>

<span class='kw'>sq</span> <span class='op'>&lt;-</span> <span class='fu'><a href='construct_sq.html'>construct_sq</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"TCTTTGGTTATCTAGCTGTATGA"</span>, <span class='st'>"CTTTGGTTATCTAGCTGTATGA"</span>, 
                     <span class='st'>"TCTTTGGTTATCTAGCTGTATG"</span>, <span class='st'>"TCTTTGGTTATCTAGCTGTATGAA"</span>, 
                     <span class='st'>"ACTGCTG"</span>, <span class='st'>"CTTAGA"</span>, <span class='st'>"CCCT"</span>, <span class='st'>"CTGAATGT"</span>), type = <span class='st'>"dna"</span>)

<span class='co'># Get an overview of the sequences and show the first four only:</span>
<span class='co'># The first four isomiRs sequences have lengths of 15, 14, 14 and 15</span>
<span class='co'># nucleotides. The remaining artificial sequences have lengths of 5, 4, 3 </span>
<span class='co'># and 5 nucleotides, respectively.</span>
<span class='fu'><a href='https://rdrr.io/r/base/summary.html'>summary</a></span>(<span class='kw'>sq</span>)</div><div class='output co'>#&gt;      Length Class  Mode
#&gt; [1,] 9      -none- raw 
#&gt; [2,] 9      -none- raw 
#&gt; [3,] 9      -none- raw 
#&gt; [4,] 9      -none- raw 
#&gt; [5,] 3      -none- raw 
#&gt; [6,] 3      -none- raw 
#&gt; [7,] 2      -none- raw 
#&gt; [8,] 3      -none- raw </div><div class='input'>
<span class='co'># Working with the first four miRNA/isomiR</span>
<span class='co'># Removing first letter from the first four sequences (miRNAs/isomiRs).</span>
<span class='co'># The sequence 1, 3 and 4 appear to be more similar.</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>[<span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span>], <span class='op'>-</span><span class='fl'>1</span>)</div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] CTTTGGTTATCTAGCTGTATGA                                                  &lt;22&gt;
#&gt; [2] TTTGGTTATCTAGCTGTATGA                                                   &lt;21&gt;
#&gt; [3] CTTTGGTTATCTAGCTGTATG                                                   &lt;21&gt;
#&gt; [4] CTTTGGTTATCTAGCTGTATGAA                                                 &lt;23&gt;</div><div class='input'>
<span class='co'># Extracting first five letters from each miRNA/ismomiR sequence:</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>[<span class='fl'>1</span><span class='op'>:</span><span class='fl'>4</span>], <span class='fl'>1</span><span class='op'>:</span><span class='fl'>5</span>)</div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] TCTTT                                                                    &lt;5&gt;
#&gt; [2] CTTTG                                                                    &lt;5&gt;
#&gt; [3] TCTTT                                                                    &lt;5&gt;
#&gt; [4] TCTTT                                                                    &lt;5&gt;</div><div class='input'>
<span class='co'># Working with all sequences</span>
<span class='co'># Extracting first letter from each sequence:</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>, <span class='fl'>1</span>)</div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] T                                                                        &lt;1&gt;
#&gt; [2] C                                                                        &lt;1&gt;
#&gt; [3] T                                                                        &lt;1&gt;
#&gt; [4] T                                                                        &lt;1&gt;
#&gt; [5] A                                                                        &lt;1&gt;
#&gt; [6] C                                                                        &lt;1&gt;
#&gt; [7] C                                                                        &lt;1&gt;
#&gt; [8] C                                                                        &lt;1&gt;</div><div class='input'>
<span class='co'># Extracting first three letters from each sequence:</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>3</span>)</div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] TCT                                                                      &lt;3&gt;
#&gt; [2] CTT                                                                      &lt;3&gt;
#&gt; [3] TCT                                                                      &lt;3&gt;
#&gt; [4] TCT                                                                      &lt;3&gt;
#&gt; [5] ACT                                                                      &lt;3&gt;
#&gt; [6] CTT                                                                      &lt;3&gt;
#&gt; [7] CCC                                                                      &lt;3&gt;
#&gt; [8] CTG                                                                      &lt;3&gt;</div><div class='input'>
<span class='co'># Extracting second, fourth, third and second letters:</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>2</span>,<span class='fl'>4</span>,<span class='fl'>3</span>,<span class='fl'>2</span>))</div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] CTTC                                                                     &lt;4&gt;
#&gt; [2] TTTT                                                                     &lt;4&gt;
#&gt; [3] CTTC                                                                     &lt;4&gt;
#&gt; [4] CTTC                                                                     &lt;4&gt;
#&gt; [5] CGTC                                                                     &lt;4&gt;
#&gt; [6] TATT                                                                     &lt;4&gt;
#&gt; [7] CTCC                                                                     &lt;4&gt;
#&gt; [8] TAGT                                                                     &lt;4&gt;</div><div class='input'>
<span class='co'># Extracting second to fifth letter - NA introduced:</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>, <span class='fl'>2</span><span class='op'>:</span><span class='fl'>5</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: some sequences are subsetted with index bigger than length - NA introduced</span></div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] CTTT                                                                     &lt;4&gt;
#&gt; [2] TTTG                                                                     &lt;4&gt;
#&gt; [3] CTTT                                                                     &lt;4&gt;
#&gt; [4] CTTT                                                                     &lt;4&gt;
#&gt; [5] CTGC                                                                     &lt;4&gt;
#&gt; [6] TTAG                                                                     &lt;4&gt;
#&gt; [7] CCT!                                                                     &lt;4&gt;
#&gt; [8] TGAA                                                                     &lt;4&gt;</div><div class='input'>
<span class='co'># Extracting all from first to twentieth - NA introduced:</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>20</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: some sequences are subsetted with index bigger than length - NA introduced</span></div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] TCTTTGGTTATCTAGCTGTA                                                    &lt;20&gt;
#&gt; [2] CTTTGGTTATCTAGCTGTAT                                                    &lt;20&gt;
#&gt; [3] TCTTTGGTTATCTAGCTGTA                                                    &lt;20&gt;
#&gt; [4] TCTTTGGTTATCTAGCTGTA                                                    &lt;20&gt;
#&gt; [5] ACTGCTG!!!!!!!!!!!!!                                                    &lt;20&gt;
#&gt; [6] CTTAGA!!!!!!!!!!!!!!                                                    &lt;20&gt;
#&gt; [7] CCCT!!!!!!!!!!!!!!!!                                                    &lt;20&gt;
#&gt; [8] CTGAATGT!!!!!!!!!!!!                                                    &lt;20&gt;</div><div class='input'>
<span class='co'># Extracting all excluding first letter of sequence:</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>, <span class='op'>-</span><span class='fl'>1</span>)</div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] CTTTGGTTATCTAGCTGTATGA                                                  &lt;22&gt;
#&gt; [2] TTTGGTTATCTAGCTGTATGA                                                   &lt;21&gt;
#&gt; [3] CTTTGGTTATCTAGCTGTATG                                                   &lt;21&gt;
#&gt; [4] CTTTGGTTATCTAGCTGTATGAA                                                 &lt;23&gt;
#&gt; [5] CTGCTG                                                                   &lt;6&gt;
#&gt; [6] TTAGA                                                                    &lt;5&gt;
#&gt; [7] CCT                                                                      &lt;3&gt;
#&gt; [8] TGAATGT                                                                  &lt;7&gt;</div><div class='input'>
<span class='co'># Extracting all excluding second and sixth letter of sequence:</span>
<span class='fu'>bite</span>(<span class='kw'>sq</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='op'>-</span><span class='fl'>2</span>, <span class='op'>-</span><span class='fl'>6</span>))</div><div class='output co'>#&gt; dna (DNA), cln (cleaned) sequences list:
#&gt; [1] TTTTGTTATCTAGCTGTATGA                                                   &lt;21&gt;
#&gt; [2] CTTGTTATCTAGCTGTATGA                                                    &lt;20&gt;
#&gt; [3] TTTTGTTATCTAGCTGTATG                                                    &lt;20&gt;
#&gt; [4] TTTTGTTATCTAGCTGTATGAA                                                  &lt;22&gt;
#&gt; [5] ATGCG                                                                    &lt;5&gt;
#&gt; [6] CTAG                                                                     &lt;4&gt;
#&gt; [7] CCT                                                                      &lt;3&gt;
#&gt; [8] CGAAGT                                                                   &lt;6&gt;</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Michal Burdukiewicz, Dominik Rafacz, Mateusz Bakala.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


