<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Construct sq object from character vector — construct_sq • tidysq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Construct sq object from character vector — construct_sq" />

<meta property="og:description" content="This function allows the user to construct objects of 
class sq from a character vector." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidysq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/michbur/tidysq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Construct sq object from character vector</h1>
    <small class="dont-index">Source: <a href='https://github.com/michbur/tidysq/blob/master/R/sq.R'><code>R/sq.R</code></a></small>
    <div class="hidden name"><code>construct_sq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function allows the user to construct objects of 
<code><a href='sq.html'>class sq</a></code> from a character vector.</p>
    
    </div>

    <pre class="usage"><span class='fu'>construct_sq</span>(<span class='no'>sq</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>is_clean</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>non_standard</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='fu'>construct_sq_ami</span>(<span class='no'>sq</span>, <span class='kw'>is_clean</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='fu'>construct_sq_nuc</span>(<span class='no'>sq</span>, <span class='kw'>is_clean</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sq</th>
      <td><p>a <code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character</a></code> vector.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p>a <code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character</a></code> string indicating type of <code>sq</code> object that
is going to be constructed; supported values are "ami" for amino acid sequences,
"nuc" for nucleotide sequences, "unt" for and <code>NULL</code> for type guessing (see details)</p></td>
    </tr>
    <tr>
      <th>is_clean</th>
      <td><p>a <code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical</a></code> value indicating if sequences are clean.
or in other words - they don't contain ambiguous values; supported values are <code>TRUE</code> 
for clean sequences, <code>FALSE</code> for unclean sequences and <code>NULL</code> for auto detecting
(see details).</p></td>
    </tr>
    <tr>
      <th>non_standard</th>
      <td><p>a <code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character</a></code> vector indicating non-standard letters
contained in sequences. If <code>NULL</code>, sequences will not be searched for non-standard letters
of length more than one. Each element of this parameter should be at least two characters 
long.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>object of <code><a href='sq.html'>class sq</a></code> with appropriate type (one of: <strong>ami</strong>,
<strong>nuc</strong>, <strong>unt</strong>, <strong>atp</strong>).</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Function <code>construct_sq</code> covers all possibilities of standard and non-standard types and 
alphabets. You can check what 'type' and 'alphabet' exactly are in <code><a href='sq.html'>sq class</a></code> 
documentation. Below there is a guide how function operates and how the program behaves 
according to the given arguments and the letters in the sequences.</p>
<p>Functions <code>construct_sq_ami</code> and <code>construct_sq_nuc</code> are wrappers around 
<code>construct_sq</code> with specified <code>type</code> parameter - accordingly "ami" or "nuc". You
can also pass "is_clean" parameter to those functions, but you cannot pass "non_standard".</p>
<p><code>sq</code> parameter should be a character vector. Each element of this vector is a biological 
sequence. If this parameter has length 0, object of class <code>sq</code> with 0 sequences will be 
created (if not specified, it will have <strong>nuc</strong> <strong>cln</strong> type, which is a result of 
rules written below). If it contains sequences of length 0, <code><a href='sq.html'>NULL</a></code> sequences
will be introduced (see <em>NULL (empty) sequences</em> section in <code><a href='sq.html'>sq class</a></code>).</p>
<p><strong>Important note:</strong> in all below cases word 'letter' stands for an element of an alphabet.
Letter might consist of more than one character, for example "Ala" might be a single letter.
However, if you want to construct or read sequences with multi-character letters, one has 
to specify <code>non_standard</code> parameter. Details of letters, alphabet and types can be 
found in <code><a href='sq.html'>sq class</a></code> documentation.</p>
    
    <h2 class="hasAnchor" id="simple-guide-to-construct-"><a class="anchor" href="#simple-guide-to-construct-"></a>Simple guide to construct </h2>

    
    <p>In most cases, just the <code>sq</code> parameter needs to be specified - type of sequences
will be guessed accordingly to rules described below. You need to pay attention, however, 
because for short sequences type may be guessed incorrectly - in this case you should
specify <code>type</code> and/or <code>is_clean</code>.</p>
<p>If your sequences contain non-standard letters, where each non-standard letter is one
character long, you also don't need to specify any parameter. Optionally, you can explicitly
do it by setting <code>type</code> to "unt".</p>
<p>If you want to construct sequences with multicharacter letters, you have to specify 
<code>non_standard</code> parameter, where you have to provide all non-standard letters longer
than one character.</p>
<p>In <code><a href='fast-mode.html'>fast mode</a></code> you have to specify both <code>type</code> and <code>is_clean</code> 
parameters. You cannot specify <code>non_standard</code> parameter in this mode. All letters
outside specified alphabet will be red as <code>NA</code> values.</p>
    
    <h2 class="hasAnchor" id="detailed-guide-to-construct-"><a class="anchor" href="#detailed-guide-to-construct-"></a>Detailed guide to construct </h2>

    
    <p>Below all possibilities that can occur during the construction of a <code>sq</code> object are listed.</p>
<p>In normal mode (no <code><a href='fast-mode.html'>fast mode</a></code>):</p><ul>
<li><p>If you don't specify any other parameter than <code>sq</code>, function will try to guess
sequence type and if it is clean (it will check in exactly this order):</p><ol>
<li><p>If it contains only ACGTU- letters, either lowercase or 
uppercase, type will be set to <strong>nuc</strong> <strong>cln</strong>.</p></li>
<li><p>If it contains any letters from 1. and additionally letters DEFHIKLMNPQRSVWY*, either
lowercase or uppercase, type will be set to <strong>ami</strong> <strong>cln</strong>.</p></li>
<li><p>If it contains any letters from 1. and additionally letters WSMKRYBDHVN, either
lowercase or uppercase, and does not contain any other letter, type will be set to 
<strong>nuc</strong> without <strong>cln</strong> subtype.</p></li>
<li><p>If it contains any letters from 1., 2., 3. and additionally letters JOUXZ, type will
be set to <strong>ami</strong> without <strong>cln</strong> subtype.</p></li>
<li><p>If it contains any letters that exceed all groups mentioned above, type will be set
to "unt".</p></li>
</ol></li>
<li><p>If you specify <code>type</code> parameter as "ami" or "nuc" (and do not specify neither 
<code>is_clean</code> nor <code>non_standard</code>) type will be checked during construction:</p><ol>
<li><p>If all letters in sequences fit the clean alphabet of given type, type will be set to 
given type with <strong>cln</strong> subtype.</p></li>
<li><p>If all letters in sequences fit the unclean alphabet of given type, type will be set to 
given type without <strong>cln</strong> subtype.</p></li>
<li><p>If at least one sequence contains at least one letter that is not an element of an  
unclean alphabet of the provided type, an error will be thrown.</p></li>
</ol></li>
<li><p>If you specify both <code>type</code> and <code>is_clean</code>, function checks if letters
in sequences matches exactly specified alphabet (with capitalisation accuracy). If they do, 
type will be set to it. Otherwise, an error will be thrown.</p></li>
<li><p>If you specify <code>type</code> as "unt" and won't neither <code>is_clean</code> nor 
<code>non_standard</code>, type will be set to <strong>unt</strong>. Letters won't be converted to uppercase, 
alphabet will consist of all letters found in sequences.</p></li>
<li><p>If you do not specify neither <code>type</code> nor <code>is_clean</code> and specify 
<code>non_standard</code> parameter, which should be character vector where each element is at least 
two characters long, all strings as specified will be detected in sequences and treated as 
letters in constructed <strong>atp</strong> <code>sq</code>.</p></li>
<li><p>All other combinations of parameters are incorrect.</p></li>
</ul>
    <p>In <code><a href='fast-mode.html'>fast mode</a></code> you have to specify <code>type</code> (it has to have either
"ami" or "nuc" value) and <code>is_clean</code> (<code>TRUE</code> or <code>FALSE</code>). You cannot specify
<code>non_standard</code>. All letters that aren't elements of destination alphabet (with a letter 
size accuracy) will be treated as <code>NA</code> values.</p>
    
    <h2 class="hasAnchor" id="handling-with-atp-and-unt-sq-and-na-values"><a class="anchor" href="#handling-with-atp-and-unt-sq-and-na-values"></a>Handling with atp and unt sq and NA values</h2>

    
    <p>You can convert letters into another using <code><a href='substitute_letters.html'>substitute_letters</a></code> and then you
can use <code><a href='typify.html'>typify</a></code> function to set of <code>sq</code> to <strong>ami</strong> or <strong>nuc</strong>.
If your sequences contain <code>NA</code> values, use <code><a href='remove_na.html'>remove_na</a></code></p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='sq.html'>sq</a></code> <code><a href='read_fasta.html'>read_fasta</a></code> <code><a href='tidysq-options.html'>tidysq-options</a></code> 
<code><a href='fast-mode.html'>fast-mode</a></code> <code><a href='substitute_letters.html'>substitute_letters</a></code> <code><a href='remove_na.html'>remove_na</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># saving option:</span>
<span class='no'>previous_option</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/options'>getOption</a></span>(<span class='st'>"tidysq_fast_mode"</span>)

<span class='co'>#### constructing sq in normal mode:</span>
<span class='co'>## setting an option:</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/options'>options</a></span>(<span class='kw'>tidysq_fast_mode</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'>## constructing sq without specifying type</span>
<span class='co'># nuc cln sq</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ATGC"</span>, <span class='st'>"TCGTTA"</span>, <span class='st'>"TT--AG"</span>))</div><div class='output co'>#&gt; nuc (nucleotides), cln (cleaned) sequences list:
#&gt; [1] <span style='color: #00BB00;'>ATGC</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>TCGTTA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>TT--AG</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># ami cln sq</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"YQQPAVVM"</span>, <span class='st'>"PQCFL"</span>))</div><div class='output co'>#&gt; ami (amino acids), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>YQQPAVVM</span><span>                                                                 </span><span style='color: #0000BB;'>&lt;8&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>PQCFL</span><span>                                                                    </span><span style='color: #0000BB;'>&lt;5&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># ami cln sq can contain * - letter meaning end of translation:</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"MMDF*"</span>, <span class='st'>"SYIHR*"</span>, <span class='st'>"MGG*"</span>))</div><div class='output co'>#&gt; ami (amino acids), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>MMDF*</span><span>                                                                    </span><span style='color: #0000BB;'>&lt;5&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>SYIHR*</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>MGG*</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># nuc sq</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"WHDHKYN"</span>, <span class='st'>"GCYVCYU"</span>))</div><div class='output co'>#&gt; nuc (nucleotides) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>WHDHKYN</span><span>                                                                  </span><span style='color: #0000BB;'>&lt;7&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>GCYVCYU</span><span>                                                                  </span><span style='color: #0000BB;'>&lt;7&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># ami sq</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"XYOQWWKCNJLO"</span>))</div><div class='output co'>#&gt; ami (amino acids) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>XYOQWWKCNJLO</span><span>                                                            </span><span style='color: #0000BB;'>&lt;12&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># unt sq - let's assume that one wants to mark some special element in sequence with %</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"%%YAPLAA"</span>, <span class='st'>"PLAA"</span>))</div><div class='output co'>#&gt; unt (unspecified type) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>%%YAPLAA</span><span>                                                                 </span><span style='color: #0000BB;'>&lt;8&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>PLAA</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'>## constructing sq with type </span>
<span class='co'># all above examples will result in an identical if specified type as guessed</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ATGC"</span>, <span class='st'>"TCGTTA"</span>, <span class='st'>"TT--AG"</span>), <span class='st'>"nuc"</span>)</div><div class='output co'>#&gt; nuc (nucleotides), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>ATGC</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>TCGTTA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>TT--AG</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"YQQPAVVM"</span>, <span class='st'>"PQCFL"</span>), <span class='st'>"ami"</span>)</div><div class='output co'>#&gt; ami (amino acids), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>YQQPAVVM</span><span>                                                                 </span><span style='color: #0000BB;'>&lt;8&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>PQCFL</span><span>                                                                    </span><span style='color: #0000BB;'>&lt;5&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"MMDF*"</span>, <span class='st'>"SYIHR*"</span>, <span class='st'>"MGG*"</span>), <span class='st'>"ami"</span>)</div><div class='output co'>#&gt; ami (amino acids), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>MMDF*</span><span>                                                                    </span><span style='color: #0000BB;'>&lt;5&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>SYIHR*</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>MGG*</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"WHDHKYN"</span>, <span class='st'>"GCYVCYU"</span>), <span class='st'>"nuc"</span>)</div><div class='output co'>#&gt; nuc (nucleotides) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>WHDHKYN</span><span>                                                                  </span><span style='color: #0000BB;'>&lt;7&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>GCYVCYU</span><span>                                                                  </span><span style='color: #0000BB;'>&lt;7&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"XYOQWWKCNJLO"</span>), <span class='st'>"ami"</span>)</div><div class='output co'>#&gt; ami (amino acids) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>XYOQWWKCNJLO</span><span>                                                            </span><span style='color: #0000BB;'>&lt;12&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"%%YAPLAA"</span>, <span class='st'>"PLAA"</span>), <span class='st'>"unt"</span>)</div><div class='output co'>#&gt; unt (unspecified type) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>%%YAPLAA</span><span>                                                                 </span><span style='color: #0000BB;'>&lt;8&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>PLAA</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># you can also use wrappers instead of parameters</span>
<span class='fu'>construct_sq_nuc</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ATGC"</span>, <span class='st'>"TCGTTA"</span>, <span class='st'>"TT--AG"</span>))</div><div class='output co'>#&gt; nuc (nucleotides), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>ATGC</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>TCGTTA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>TT--AG</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq_ami</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"YQQPAVVM"</span>, <span class='st'>"PQCFL"</span>))</div><div class='output co'>#&gt; ami (amino acids), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>YQQPAVVM</span><span>                                                                 </span><span style='color: #0000BB;'>&lt;8&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>PQCFL</span><span>                                                                    </span><span style='color: #0000BB;'>&lt;5&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq_ami</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"MMDF*"</span>, <span class='st'>"SYIHR*"</span>, <span class='st'>"MGG*"</span>))</div><div class='output co'>#&gt; ami (amino acids), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>MMDF*</span><span>                                                                    </span><span style='color: #0000BB;'>&lt;5&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>SYIHR*</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>MGG*</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq_nuc</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"WHDHKYN"</span>, <span class='st'>"GCYVCYU"</span>))</div><div class='output co'>#&gt; nuc (nucleotides) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>WHDHKYN</span><span>                                                                  </span><span style='color: #0000BB;'>&lt;7&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>GCYVCYU</span><span>                                                                  </span><span style='color: #0000BB;'>&lt;7&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq_ami</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"XYOQWWKCNJLO"</span>))</div><div class='output co'>#&gt; ami (amino acids) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>XYOQWWKCNJLO</span><span>                                                            </span><span style='color: #0000BB;'>&lt;12&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># One can force type other than guessed (if letters fit in the destination alphabet)</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ATGC"</span>, <span class='st'>"TCGTTA"</span>, <span class='st'>"TT--AG"</span>), <span class='st'>"nuc"</span>, <span class='kw'>is_clean</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt; nuc (nucleotides) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>ATGC</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>TCGTTA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>TT--AG</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ATGC"</span>, <span class='st'>"TCGTTA"</span>, <span class='st'>"TT--AG"</span>), <span class='st'>"ami"</span>)</div><div class='output co'>#&gt; ami (amino acids), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>ATGC</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>TCGTTA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>TT--AG</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ATGC"</span>, <span class='st'>"TCGTTA"</span>, <span class='st'>"TT--AG"</span>), <span class='st'>"ami"</span>, <span class='kw'>is_clean</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt; ami (amino acids) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>ATGC</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>TCGTTA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>TT--AG</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ATGC"</span>, <span class='st'>"TCGTTA"</span>, <span class='st'>"TT--AG"</span>), <span class='st'>"unt"</span>)</div><div class='output co'>#&gt; unt (unspecified type) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>ATGC</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>TCGTTA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>TT--AG</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'>## constructing with non_standard specified</span>
<span class='co'># in sequences below "mA" denotes methyled alanine - two characters are treated as single letter</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"LmAQYmASSR"</span>, <span class='st'>"LmASMKLKFmAmA"</span>), <span class='kw'>non_standard</span> <span class='kw'>=</span> <span class='st'>"mA"</span>)</div><div class='output co'>#&gt; atp (atypical alphabet) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>L mA Q Y mA S S R</span><span>                                                        </span><span style='color: #0000BB;'>&lt;8&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>L mA S M K L K F mA mA</span><span>                                                  </span><span style='color: #0000BB;'>&lt;10&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># reading sequences with three-letter names:</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"ProProGlyAlaMetAlaCys"</span>), <span class='kw'>non_standard</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"Pro"</span>, <span class='st'>"Gly"</span>, <span class='st'>"Ala"</span>, <span class='st'>"Met"</span>, <span class='st'>"Cys"</span>))</div><div class='output co'>#&gt; atp (atypical alphabet) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>Pro Pro Gly Ala Met Ala Cys</span><span>                                              </span><span style='color: #0000BB;'>&lt;7&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'>#### constructing in fast mode:</span>
<span class='co'>## setting fast mode on</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/options'>options</a></span>(<span class='kw'>tidysq_fast_mode</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># you cannot construct without specifying type</span></div></span><span class='co'># NOT RUN {</span>
<span class='fu'>construct_sq</span>(<span class='st'>"CTGA"</span>)
<span class='co'># }</span><div class='input'><span class='fu'>construct_sq</span>(<span class='st'>"CTGA"</span>, <span class='st'>"nuc"</span>, <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; nuc (nucleotides), cln (cleaned) sequences list:
#&gt; [1] <span style='color: #00BB00;'>CTGA</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; </div><div class='input'>
# you cannot construct with specifying non_standard
</div></span><span class='co'># NOT RUN {</span>
<span class='fu'>construct_sq</span>(<span class='st'>"mAPQ"</span>, <span class='kw'>non_standard</span> <span class='kw'>=</span> <span class='st'>"mA"</span>)
<span class='co'># }</span><div class='input'>
<span class='co'># letters other than in specified alphabet will be treated as NA:</span>
<span class='fu'>construct_sq</span>(<span class='st'>"NCTGCNA"</span>, <span class='st'>"nuc"</span>, <span class='fl'>TRUE</span>)</div><div class='output co'>#&gt; nuc (nucleotides), cln (cleaned) sequences list:
#&gt; [1] <span style='color: #555555;'>!</span><span style='color: #00BB00;'>CTGC</span><span style='color: #555555;'>!</span><span style='color: #00BB00;'>A</span><span>                                                                  </span><span style='color: #0000BB;'>&lt;7&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'>#### Other examples:</span>
<span class='co'>## setting fast mode off again:</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/options'>options</a></span>(<span class='kw'>tidysq_fast_mode</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># lowercase letters are converted to uppercase if detected type is "ami" or "nuc"</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"aTGc"</span>, <span class='st'>"tcgTTA"</span>, <span class='st'>"tt--AG"</span>))</div><div class='output co'>#&gt; nuc (nucleotides), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>ATGC</span><span>                                                                     </span><span style='color: #0000BB;'>&lt;4&gt;</span><span>
#&gt; [2] </span><span style='color: #00BB00;'>TCGTTA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>TT--AG</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'><span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"XYOqwwKCNJLo"</span>))</div><div class='output co'>#&gt; ami (amino acids) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>XYOQWWKCNJLO</span><span>                                                            </span><span style='color: #0000BB;'>&lt;12&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># but not for "unt"</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"aAAaAA"</span>), <span class='kw'>type</span> <span class='kw'>=</span> <span class='st'>"unt"</span>)</div><div class='output co'>#&gt; unt (unspecified type) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>aAAaAA</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'># you can construct sq with length 0</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/character'>character</a></span>(<span class='fl'>0</span>))</div><div class='output co'>#&gt; nuc (nucleotides), cln (cleaned) sequences list of length 0</div><div class='input'>
<span class='co'># and sq with empty sequences</span>
<span class='fu'>construct_sq</span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"AGTGGC"</span>, <span class='st'>""</span>, <span class='st'>"CATGA"</span>, <span class='st'>""</span>))</div><div class='output co'>#&gt; nuc (nucleotides), cln (cleaned) sequences list:
#&gt; [1] </span><span style='color: #00BB00;'>AGTGGC</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;6&gt;</span><span>
#&gt; [2] </span><span style='color: #555555;'>&lt;NULL&gt;</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;0&gt;</span><span>
#&gt; [3] </span><span style='color: #00BB00;'>CATGA</span><span>                                                                    </span><span style='color: #0000BB;'>&lt;5&gt;</span><span>
#&gt; [4] </span><span style='color: #555555;'>&lt;NULL&gt;</span><span>                                                                   </span><span style='color: #0000BB;'>&lt;0&gt;</span><span>
#&gt; </div><div class='input'>
<span class='co'>## reseting an option</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/options'>options</a></span>(<span class='kw'>tidysq_fast_mode</span> <span class='kw'>=</span> <span class='no'>previous_option</span>)</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#simple-guide-to-construct-">Simple guide to construct </a></li>

      <li><a href="#detailed-guide-to-construct-">Detailed guide to construct </a></li>

      <li><a href="#handling-with-atp-and-unt-sq-and-na-values">Handling with atp and unt sq and NA values</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michal Burdukiewicz, Dominik Rafacz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

