<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>sq: class for keeping biological sequences tidy — sq • tidysq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="sq: class for keeping biological sequences tidy — sq" />

<meta property="og:description" content="An object of class sq represents a list of biological sequences. It is the 
main internal format of the tidysq package and most functions operate on it. 
The storage method is memory-optimized so that objects require as little memory
as possible (details below)." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidysq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/michbur/tidysq">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>sq: class for keeping biological sequences tidy</h1>
    <small class="dont-index">Source: <a href='https://github.com/michbur/tidysq/blob/master/R/sq.R'><code>R/sq.R</code></a></small>
    <div class="hidden name"><code>sq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>An object of class <strong>sq</strong> represents a list of biological sequences. It is the 
main internal format of the <span class="pkg">tidysq</span> package and most functions operate on it. 
The storage method is memory-optimized so that objects require as little memory
as possible (details below).</p>
    
    </div>

        
    <h2 class="hasAnchor" id="construction-reading-import-of-sq-objects"><a class="anchor" href="#construction-reading-import-of-sq-objects"></a>Construction/reading/import of sq objects</h2>

    
    <p>There are multiple ways of obtaining <code>sq</code> objects:</p><ul>
<li><p>constructing from a <code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character</a></code> vector with <code><a href='construct_sq.html'>construct_sq</a></code>,</p></li>
<li><p>constructing from another object with <code><a href='as.sq.html'>as.sq</a></code> method,</p></li>
<li><p>reading from the fasta file with <code><a href='read_fasta.html'>read_fasta</a></code>,</p></li>
<li><p>importing from a format of other package like <span class="pkg">ape</span> or
<span class="pkg">Biostrings</span> with <code><a href='import_sq.html'>import_sq</a></code>.</p></li>
</ul>
    <p><strong>Important note:</strong> A manual assignment of a class <code>sq</code> to an object is
<strong>strongly discouraged</strong> - due to the usage of low-level functions for
bit packing such assignment may lead to calling one of those functions during
operating on object or even printing it which can cause a crash of R session and,
in consequence, loss of data.</p>
    
    <h2 class="hasAnchor" id="export-writing-of-sq-objects"><a class="anchor" href="#export-writing-of-sq-objects"></a>Export/writing of sq objects</h2>

    
    <p>There are multiple ways of saving <code>sq</code> objects or converting them into
other formats:</p><ul>
<li><p>converting into a character vector with <code><a href='as.character.sq.html'>as.character</a></code> 
method,</p></li>
<li><p>converting into a character matrix (or numeric matrix, in case of <strong>enc</strong>) 
with <code><a href='as.matrix.sq.html'>as.matrix</a></code> method,</p></li>
<li><p>converting into a list of numerics with <code><a href='encsq_to_list.html'>encsq_to_list</a></code> (only
for <strong>enc</strong> <code>sq</code>),</p></li>
<li><p>saving as fasta file with <code><a href='write_fasta.html'>write_fasta</a></code>,</p></li>
<li><p>exporting into a format of other package like <code>ape</code> or
<code>Biostrings</code> with <code><a href='export_sq.html'>export_sq</a></code>.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="types-of-sq"><a class="anchor" href="#types-of-sq"></a>Types of sq</h2>

    
    <p>This package is meant to handle both amino acids and nucleotides sequences 
thus there is need to differentiate <code>sq</code> objects that keep them. In 
addition, there are special types for handling non-standard sequence 
formats and encodings.</p>
<p>Each <strong>sq</strong> object has exactly one of <strong>types</strong>:</p><ul>
<li><p><strong>ami</strong> - (<em>amino acids</em>) represents a list of sequences of amino acids
(peptides or proteins),</p></li>
<li><p><strong>nuc</strong> - (<em>nucleotides</em>) represents a list of nucleotide sequences (RNA or 
DNA),</p></li>
<li><p><strong>unt</strong> - (<em>untyped</em>) represents a list of sequences that do not have 
specified type. They are mainly result of reading sequences from a file that 
contains some letters that are not in standard nucleotide or amino acid alphabets 
and user has not specified them explicitly. They should be converted to <strong>ami</strong> 
or <strong>nuc</strong> sequences (using functions like <code><a href='substitute_letters.html'>substitute_letters</a></code> or 
<code><a href='typify.html'>typify</a></code>).</p></li>
<li><p><strong>atp</strong> - (<em>atypical</em>) represents sequences that have an alphabet
different from standard <strong>ami</strong> or <strong>nuc</strong> alphabets - similarly to 
<strong>unt</strong>, but user has explicitly informed about it. They are
result of constructing sequences or reading from file with specifying 
<code>non_standard</code> parameter (for details see <code><a href='read_fasta.html'>read_fasta</a></code>
and <code><a href='construct_sq.html'>construct_sq</a></code>). They are also result of using function
<code><a href='substitute_letters.html'>substitute_letters</a></code> - users can use this for example to simplify 
an alphabet and to replace a few letters by one.</p></li>
<li><p><strong>enc</strong> - (<em>encoded</em>) represents list of sequences that have been
encoded with function <code><a href='encode.html'>encode</a></code> where each letter is assigned with 
numeric value.</p></li>
</ul>
    <p>Additionally, there is a special subtype <strong>cln</strong> (standing for <em>clean</em>).
Only <strong>ami</strong> and <strong>nuc</strong> <code>sq</code> objects may have this subtype. It indicates
that sequences do not contain ambiguous letters (see "alphabets" section below).</p>
<p><code>sq</code> object type is printed when using overloaded method 
<code><a href='print.sq.html'>print</a></code>. It can be also checked by using <code><a href='get_sq_type.html'>get_sq_type</a></code></p>
    
    <h2 class="hasAnchor" id="alphabet"><a class="anchor" href="#alphabet"></a>Alphabet</h2>

    
    <p>Each <code>sq</code> object have an <strong>alphabet</strong> associated with it. Alphabet is
a set of possible <strong>letters</strong> that can appear in sequences contained in object.
Alphabet is kept mostly as a character vector, where each element represents one
<strong>letter</strong>.</p>
<p><code>sq</code> objects of type <strong>ami</strong> or <strong>nuc</strong> have fixed alphabets (depending
also if object has <strong>cln</strong> subtype). In other words, if two <code>sq</code> objects have
exactly the same type - <strong>ami</strong> or <strong>nuc</strong> - and either both have or both do
not have a <strong>cln</strong> subtype, they are ensured to have the same alphabets.</p>
<p>Here are listed alphabets for these types:</p><ul>
<li><p><strong>ami</strong> <strong>cln</strong> - ACDEFGHIKLMNPQRSTVWY-*</p></li>
<li><p><strong>ami</strong>, (not <strong>cln</strong>) - ABCDEFGHIJKLMNOPQRSTUVWXYZ-*</p></li>
<li><p><strong>nuc</strong> <strong>cln</strong> - ACGTU-</p></li>
<li><p><strong>nuc</strong>, (not <strong>cln</strong>) - ACGTUWSMKRYBDHVN-</p></li>
</ul>
    <p>To see details of these alphabets see <code><a href='aminoacids_df.html'>aminoacids_df</a></code> and 
<code><a href='nucleotides_df.html'>nucleotides_df</a></code>.</p>
<p>Other types of <code>sq</code> objects are allowed to have different alphabets. Having an alphabet
exactly identical to one of those above does not automatically indicate that the type of
the sequence is one of those - e. g., there might be <strong>unt</strong> <code>sq</code> that has an alphabet
identical to <strong>ami</strong> <strong>cln</strong> alphabet. To set the type, one should use the
<code><a href='typify.html'>typify</a></code> function.</p>
<p>The purpose of co-existence of <strong>unt</strong> and <strong>atp</strong> alphabets is the 
fact that although there is a standard for format of <em>fasta</em> files, sometimes 
there are other types of symbols, which do not match the standard. Thanks to these types, 
tidysq can import files with customized alphabets can be imported. Moreover, a user 
may want to group amino acids with similar properties (e. g., for machine learning) 
and replace the longer alphabet with symbols of groups. To check details, see 
<code><a href='read_fasta.html'>read_fasta</a></code>, <code><a href='construct_sq.html'>construct_sq</a></code> and <code><a href='substitute_letters.html'>substitute_letters</a></code>.</p>
<p>All of the types: <strong>ami</strong>, <strong>nuc</strong>, <strong>atp</strong>, <strong>unt</strong> have alphabets
that are character vectors while <strong>enc</strong> objects have alphabets that are numeric
vectors. They are result of <code><a href='encode.html'>encode</a></code> function, see its manual for details.</p>
<p><strong>Important note:</strong> in <strong>atp</strong> alphabets there is possibility of appearance of
letters that consists of more than one character - this functionality is provided in
order to handle situations like post-translational modifications, (e.g., using "mA" to 
indicating methylated alanine).</p>
<p><strong>Important note:</strong> alphabets of <strong>atp</strong> and <strong>unt</strong> <code>sq</code> objects
are case sensitive. Thus, in their alphabets there can appear both lowercase and
uppercase letters simultaneously and they are treated as different characters. Alphabets of
<strong>nuc</strong> and <strong>ami</strong> objects are always uppercase and all functions converts
other parameters to uppercase when working with <strong>ami</strong> or <strong>nuc</strong> - e. g.,
<code><a href='grapes-has-grapes.html'>%has%</a></code> operator converts lower letters to upper when searching for motifs in
<strong>ami</strong> or <strong>nuc</strong> object.</p>
<p><strong>Important note:</strong> maximum length of an alphabet is <strong>30 letters</strong>. You are
not allowed to read fasta files or construct from character vectors that have more
than 30 distinct characters in sequences (with exception of reading or constructing
<strong>ami</strong> or <strong>nuc</strong> objects - during their construction lowercase letters
are automatically converted to uppercase).</p>
<p>You can obtain an alphabet of the <code>sq</code> object using the <code><a href='get_sq_alphabet.html'>get_sq_alphabet</a></code>
function. You can
check which letters are invalid (are not represented in standard amino acids or nucleotides 
alphabet)
in each sequence of given <code>sq</code> object of type <strong>unt</strong> or <strong>atp</strong> by using
<code><a href='get_invalid_letters.html'>get_invalid_letters</a></code>. You can substitute one letter with another using
<code><a href='substitute_letters.html'>substitute_letters</a></code>.</p>
    
    <h2 class="hasAnchor" id="missing-not-available-values"><a class="anchor" href="#missing-not-available-values"></a>Missing/Not Available values</h2>

    
    <p>There is a possibility of introducing <code><a href='https://www.rdocumentation.org/packages/base/topics/NA'>NA</a></code> values into sequences. <code>NA</code> value
does not represents gap (which are represented by <code>'-'</code>) or wildcard elements 
(<code>N</code> in he case of nucleotides and <code>'X'</code> in the case of amino acids), but is used 
as a representation of an empty position or invalid letters (not represented in nucleotide or 
amino acid alphabet).
<code>NA</code> does not belong to any alphabet (with exception of <strong>enc</strong> objects where some of
letters might be <code>NA</code>, see <code><a href='encode.html'>encode</a></code> for details). It is printed as
'!' and, thus, it is highly unrecommended to use '!' as special letter in <strong>atp</strong>
sequences (but print character can be changed in options, see <code><a href='tidysq-options.html'>tidysq-options</a></code>).</p>
<p><code>NA</code> might be introduced by:</p><ul>
<li><p>reading fasta file with non-standard letters in
<code><a href='fast-mode.html'>fast mode</a></code> with <code><a href='read_fasta.html'>read_fasta</a></code>,</p></li>
<li><p>replacing a letter with <code>NA</code> value with <code><a href='substitute_letters.html'>substitute_letters</a></code>,</p></li>
<li><p>subsetting sequences out of their lengths with <code><a href='bite.html'>bite</a></code>.</p></li>
</ul>
    <p>An user can convert sequences that contain <code>NA</code> values into <code>NULL</code>
sequences with <code><a href='remove_na.html'>remove_na</a></code>.</p>
    
    <h2 class="hasAnchor" id="null-empty-sequences"><a class="anchor" href="#null-empty-sequences"></a>NULL (empty) sequences</h2>

    
    <p><code>NULL</code> sequence is a sequence of length 0.</p>
<p><code>NULL</code> sequences might be introduced by:</p><ul>
<li><p>constructing <code>sq</code> object from character string of length zero with
<code><a href='construct_sq.html'>construct_sq</a></code>,</p></li>
<li><p>using the <code><a href='clean.html'>clean</a></code> function,</p></li>
<li><p>using the <code><a href='remove_na.html'>remove_na</a></code> function,</p></li>
<li><p>subsetting <code>sq</code> object with <a href='sqextract.html'>extract operator</a></p></li>
</ul>
    
    <h2 class="hasAnchor" id="storage-format"><a class="anchor" href="#storage-format"></a>Storage format</h2>

    
    <p><code>sq</code> object is, in fact, <strong>list of raw vectors</strong>. The fact that it is list
implies that an user can concatenate <code>sq</code> objects using <code><a href='c.sq.html'>c</a></code> method
and subset them using <code><a href='sqextract.html'>extract operator</a></code>. Alphabet is kept
as an attribute of the object.</p>
<p>Raw vectors are the most efficient way of storage - each letter of the sequence has assigned
an integer (its index in alphabet of <code>sq</code> object). Those integers in binary format
fit in less than 8 bits, but normally are stored on 16 bits. However, thanks to bit
packing it is possible to remove unused bits and store numbers more tightly. This 
operations result in a little time overhead in all operations, because most of them 
require unpacking and repacking sequences, but this cost is relatively low in comparison
to amount of saved memory.</p>
<p>For example - <strong>nuc</strong> <strong>cln</strong> alphabet consist of 6 values: ACGTU-. They are 
assigned numbers 1 to 6 respectively. Those numbers in binary format take form: <code>001</code>,
<code>010</code>, <code>011</code>, <code>100</code>, <code>101</code>, <code>110</code>. Each of the letters can 
be coded with just 3 bits instead of 8 which is demanded by <code>char</code> - this allows
us to save more than 60% of memory spent on storage of nucleotides sequences.</p>
    
    <h2 class="hasAnchor" id="tibble-compatibility"><a class="anchor" href="#tibble-compatibility"></a>tibble compatibility</h2>

    
    <p><code>sq</code> objects are compatible with <code><a href='https://tibble.tidyverse.org/reference/tibble.html'>tibble</a></code> class - that means one can 
have a <code>sq</code> object as a column of a <code>tibble</code>. There are overloaded print methods, so
that it is printed in pretty format.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      
      <li><a href="#construction-reading-import-of-sq-objects">Construction/reading/import of sq objects</a></li>

      <li><a href="#export-writing-of-sq-objects">Export/writing of sq objects</a></li>

      <li><a href="#types-of-sq">Types of sq</a></li>

      <li><a href="#alphabet">Alphabet</a></li>

      <li><a href="#missing-not-available-values">Missing/Not Available values</a></li>

      <li><a href="#null-empty-sequences">NULL (empty) sequences</a></li>

      <li><a href="#storage-format">Storage format</a></li>

      <li><a href="#tibble-compatibility">tibble compatibility</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michal Burdukiewicz, Dominik Rafacz.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

