<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quick Start • tidysq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Quick Start">
<meta property="og:description" content="tidysq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidysq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/quick-start.html">Quick Start</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/michbur/tidysq/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="quick-start_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quick Start</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/michbur/tidysq/blob/master/vignettes/quick-start.Rmd"><code>vignettes/quick-start.Rmd</code></a></small>
      <div class="hidden name"><code>quick-start.Rmd</code></div>

    </div>

    
    
<p><code>tidysq</code> package is meant to store and conduct operations on biological sequences. This vignette provides a guide to basic usage of <code>tidysq</code>, i.e. reading, manipulating and writing sequences to file.</p>
<p>The package isn’t released on CRAN yet, so it has to be installed with <code>install_github()</code> function from <code>devtools</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># devtools::install_github("BioGenies/tidysq")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/michbur/tidysq">tidysq</a></span><span class="op">)</span></code></pre></div>
<div id="sequence-creation" class="section level1">
<h1 class="hasAnchor">
<a href="#sequence-creation" class="anchor"></a>Sequence creation</h1>
<p>Biological sequences can be and often are represented as strings – sequences of letters. For example, a DNA sequence can take the form of <code>"TAGGCCCTAGACCTG"</code>, where <code>A</code> means adenine, <code>C</code> – cytosine, <code>G</code> – guanine and <code>T</code> – thymine. Exact IUPAC recommendations for one-letter codes can be found <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC341218/">here</a>.</p>
<p>Within <code>tidysq</code> package sequence data is stored in <code>sq</code> objects, that is, vectors of biological sequences. They can be created from string vectors as above:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sq_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sq.html">sq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TAGGCCCTAGACCTG"</span>, <span class="st">"TAGGCCCTGGGCATG"</span><span class="op">)</span><span class="op">)</span>
<span class="va">sq_dna</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] TAGGCCCTAGACCTG                                                         &lt;15&gt;</span>
<span class="co">#&gt; [2] TAGGCCCTGGGCATG                                                         &lt;15&gt;</span></code></pre></div>
<p>There are several thing to note. First, each sequence is an element of <code>sq</code> object. Many operations are vectorized — they are applied to all sequences of a vector — and <code>sq</code> objects are no different in this regard. Second, the first line of output says: <code>basic DNA sequences list</code>. This means that all sequences of this object are of DNA type and do not use ambiguous letters (more about that in “Advanced alphabet techniques” vignette).</p>
</div>
<div id="subsequencing" class="section level1">
<h1 class="hasAnchor">
<a href="#subsequencing" class="anchor"></a>Subsequencing</h1>
<p>Manipulating sequence objects is an integral part of <code>tidysq</code>. <code>sq</code> objects can be easily subsetted using usual R syntax:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sq_dna</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] TAGGCCCTAGACCTG                                                         &lt;15&gt;</span></code></pre></div>
<p>Extracting subsequences is a bit more complicated than that — because it uses designated function <code><a href="../reference/bite.html">bite()</a></code>. Its syntax, however, closely resembles that of base R — indexing starts with one and negative indices are interpreted as “anything except that”. It returns an <code>sq</code> object with all sequences subsetted:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/bite.html">bite</a></span><span class="op">(</span><span class="va">sq_dna</span>, <span class="fl">5</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] CCCTAG                                                                   &lt;6&gt;</span>
<span class="co">#&gt; [2] CCCTGG                                                                   &lt;6&gt;</span>
<span class="fu"><a href="../reference/bite.html">bite</a></span><span class="op">(</span><span class="va">sq_dna</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">9</span>, <span class="op">-</span><span class="fl">11</span>, <span class="op">-</span><span class="fl">13</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] TAGGCCCTGCTG                                                            &lt;12&gt;</span>
<span class="co">#&gt; [2] TAGGCCCTGCTG                                                            &lt;12&gt;</span></code></pre></div>
<p>It’s possible to reverse sequences using this function:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Don't do it like that!</span>
<span class="fu"><a href="../reference/bite.html">bite</a></span><span class="op">(</span><span class="va">sq_dna</span>, <span class="fl">15</span><span class="op">:</span><span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] GTCCAGATCCCGGAT                                                         &lt;15&gt;</span>
<span class="co">#&gt; [2] GTACGGGTCCCGGAT                                                         &lt;15&gt;</span></code></pre></div>
<p>However, this usage is strongly discouraged, because it’s both ineffective and works badly with sequences of different lengths. Instead, there is a designated function <code><a href="../reference/reverse.html">reverse()</a></code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/reverse.html">reverse</a></span><span class="op">(</span><span class="va">sq_dna</span><span class="op">)</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] GTCCAGATCCCGGAT                                                         &lt;15&gt;</span>
<span class="co">#&gt; [2] GTACGGGTCCCGGAT                                                         &lt;15&gt;</span></code></pre></div>
<p>Note that it is very different to base <code><a href="https://rdrr.io/r/base/rev.html">rev()</a></code>, which reverses only the order of sequences, not letters:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/rev.html">rev</a></span><span class="op">(</span><span class="va">sq_dna</span><span class="op">)</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] TAGGCCCTGGGCATG                                                         &lt;15&gt;</span>
<span class="co">#&gt; [2] TAGGCCCTAGACCTG                                                         &lt;15&gt;</span></code></pre></div>
<p>We can combine two or more <code>sq</code> objects using base <code><a href="https://rdrr.io/r/base/c.html">c()</a></code> function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sq_dna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">sq_dna</span>, <span class="fu"><a href="../reference/reverse.html">reverse</a></span><span class="op">(</span><span class="va">sq_dna</span><span class="op">)</span><span class="op">)</span>
<span class="va">sq_dna</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] TAGGCCCTAGACCTG                                                         &lt;15&gt;</span>
<span class="co">#&gt; [2] TAGGCCCTGGGCATG                                                         &lt;15&gt;</span>
<span class="co">#&gt; [3] GTCCAGATCCCGGAT                                                         &lt;15&gt;</span>
<span class="co">#&gt; [4] GTACGGGTCCCGGAT                                                         &lt;15&gt;</span></code></pre></div>
</div>
<div id="biological-interpretation" class="section level1">
<h1 class="hasAnchor">
<a href="#biological-interpretation" class="anchor"></a>Biological interpretation</h1>
<p><code>tidysq</code> offers two functions specific to DNA/RNA sequences, namely <code><a href="../reference/complement.html">complement()</a></code> and <code><a href="../reference/translate.html">translate()</a></code>. The former creates sequences with complementary bases, that is, replaces <code>A</code> with <code>T</code>, <code>C</code> with <code>G</code> and <em>vice versa</em>. The latter translates input to amino acid sequences using <a href="https://en.wikipedia.org/wiki/DNA_and_RNA_codon_tables">the translation table with three-letter codons</a>.</p>
<p>These functions can be called as shown below:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/complement.html">complement</a></span><span class="op">(</span><span class="va">sq_dna</span><span class="op">)</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] ATCCGGGATCTGGAC                                                         &lt;15&gt;</span>
<span class="co">#&gt; [2] ATCCGGGACCCGTAC                                                         &lt;15&gt;</span>
<span class="co">#&gt; [3] CAGGTCTAGGGCCTA                                                         &lt;15&gt;</span>
<span class="co">#&gt; [4] CATGCCCAGGGCCTA                                                         &lt;15&gt;</span>
<span class="fu"><a href="../reference/translate.html">translate</a></span><span class="op">(</span><span class="va">sq_dna</span><span class="op">)</span>
<span class="co">#&gt; basic amino acid sequences list:</span>
<span class="co">#&gt; [1] *ALDL                                                                    &lt;5&gt;</span>
<span class="co">#&gt; [2] *ALGM                                                                    &lt;5&gt;</span>
<span class="co">#&gt; [3] VQIPD                                                                    &lt;5&gt;</span>
<span class="co">#&gt; [4] VRVPD                                                                    &lt;5&gt;</span></code></pre></div>
<p>One noteworthy feature here is that translation can be done with any genetic code table of those listed <a href="https://en.wikipedia.org/wiki/List_of_genetic_codes">on this Wikipedia page</a>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/translate.html">translate</a></span><span class="op">(</span><span class="va">sq_dna</span>, table <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>
<span class="co">#&gt; basic amino acid sequences list:</span>
<span class="co">#&gt; [1] QALDL                                                                    &lt;5&gt;</span>
<span class="co">#&gt; [2] QALGM                                                                    &lt;5&gt;</span>
<span class="co">#&gt; [3] VQIPD                                                                    &lt;5&gt;</span>
<span class="co">#&gt; [4] VRVPD                                                                    &lt;5&gt;</span></code></pre></div>
</div>
<div id="finding-motifs" class="section level1">
<h1 class="hasAnchor">
<a href="#finding-motifs" class="anchor"></a>Finding motifs</h1>
<p>Motifs are short subsequences. These are often searched for in biological sequences. <code>tidysq</code> has two distinct functions that allow the user to perform such search.</p>
<p>One of them is a <code><a href="../reference/%has%.html">%has%</a></code> operator that takes <code>sq</code> object and character vector as parameters respectively. It returns a logical vector of the same length as <code>sq</code> object, where each element says whether all motifs passed as strings were found in given sequence:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sq_dna</span> <span class="op">%has%</span> <span class="st">"ATC"</span>
<span class="co">#&gt; [1] FALSE FALSE  TRUE FALSE</span>
<span class="co"># It can be used to subset sq</span>
<span class="va">sq_dna</span><span class="op">[</span><span class="va">sq_dna</span> <span class="op">%has%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"AG"</span>, <span class="st">"CC"</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt; basic DNA sequences list:</span>
<span class="co">#&gt; [1] TAGGCCCTAGACCTG                                                         &lt;15&gt;</span>
<span class="co">#&gt; [2] TAGGCCCTGGGCATG                                                         &lt;15&gt;</span>
<span class="co">#&gt; [3] GTCCAGATCCCGGAT                                                         &lt;15&gt;</span></code></pre></div>
<p>It says nothing about motif placement within sequence nor it exact form, however. In this case, there is <code><a href="../reference/find_motifs.html">find_motifs()</a></code> function that returns a whole <code>tibble</code> (from <code>tibble</code> package; basically improved version of <code>data.frame</code>) with various info about found motifs. Important thing to note here is that the second argument is a character vector of sequence names to avoid embedding potentially long sequences in resulting <code>tibble</code> potentially many times:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/find_motifs.html">find_motifs</a></span><span class="op">(</span><span class="va">sq_dna</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"seq1"</span>, <span class="st">"seq2"</span>, <span class="st">"rev1"</span>, <span class="st">"rev2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ATC"</span>, <span class="st">"TAG"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 4 x 5</span>
<span class="co">#&gt;   names found     sought start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dna_bsc&gt; &lt;chr&gt;  &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1 rev1  ATC   &lt;3&gt; ATC        7     9</span>
<span class="co">#&gt; 2 seq1  TAG   &lt;3&gt; TAG        1     3</span>
<span class="co">#&gt; 3 seq1  TAG   &lt;3&gt; TAG        8    10</span>
<span class="co">#&gt; 4 seq2  TAG   &lt;3&gt; TAG        1     3</span></code></pre></div>
<p>There are ambiguous DNA bases in IUPAC codes and these can be used in motifs. One of them is <code>"N"</code> — its meaning is "any of <code>A</code>, <code>C</code>, <code>G</code> or <code>T</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/find_motifs.html">find_motifs</a></span><span class="op">(</span><span class="va">sq_dna</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"seq1"</span>, <span class="st">"seq2"</span>, <span class="st">"rev1"</span>, <span class="st">"rev2"</span><span class="op">)</span>, <span class="st">"GNCC"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 7 x 5</span>
<span class="co">#&gt;   names found     sought start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dna_bsc&gt; &lt;chr&gt;  &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1 seq1  GGCC  &lt;4&gt; GNCC       3     6</span>
<span class="co">#&gt; 2 seq1  GCCC  &lt;4&gt; GNCC       4     7</span>
<span class="co">#&gt; 3 seq1  GACC  &lt;4&gt; GNCC      10    13</span>
<span class="co">#&gt; 4 seq2  GGCC  &lt;4&gt; GNCC       3     6</span>
<span class="co">#&gt; 5 seq2  GCCC  &lt;4&gt; GNCC       4     7</span>
<span class="co">#&gt; 6 rev1  GTCC  &lt;4&gt; GNCC       1     4</span>
<span class="co">#&gt; 7 rev2  GTCC  &lt;4&gt; GNCC       7    10</span></code></pre></div>
<p>This example displays the difference between <code>"sought"</code> and <code>"found"</code> columns. The former contains the string representation of motif that the user was looking for, while the latter contains a <code>tidysq</code>-encoded sequence with an “instance” of motif.</p>
<p>Two additional characters are reserved because of their special meaning in motifs. <code>"^"</code> means that this motif must be found at the start of a sequence, while <code>"$"</code> means the same, but with the end instead. They can be mixed with ambiguous letters, of course:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/find_motifs.html">find_motifs</a></span><span class="op">(</span><span class="va">sq_dna</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"seq1"</span>, <span class="st">"seq2"</span>, <span class="st">"rev1"</span>, <span class="st">"rev2"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"^TAG"</span>, <span class="st">"ATN$"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 x 5</span>
<span class="co">#&gt;   names found     sought start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;dna_bsc&gt; &lt;chr&gt;  &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1 seq1  TAG   &lt;3&gt; ^TAG       1     3</span>
<span class="co">#&gt; 2 seq2  TAG   &lt;3&gt; ^TAG       1     3</span>
<span class="co">#&gt; 3 seq2  ATG   &lt;3&gt; ATN$      13    15</span></code></pre></div>
</div>
<div id="exporting-sq-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#exporting-sq-objects" class="anchor"></a>Exporting sq objects</h1>
<p>After doing computations the user might wish to save their sequences for future use. One of the most popular formats for storing biological sequences is FASTA. <code>tidysq</code> allows the user to write sequences to FASTA file with <code><a href="../reference/write_fasta.html">write_fasta()</a></code> function. Important thing to remember here is that it has to be passed <code>name</code> parameter, analogous to that present in <code><a href="../reference/find_motifs.html">find_motifs()</a></code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/write_fasta.html">write_fasta</a></span><span class="op">(</span><span class="va">sq_dna</span>,
            <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"seq1"</span>, <span class="st">"seq2"</span>, <span class="st">"rev1"</span>, <span class="st">"rev2"</span><span class="op">)</span>,
            <span class="st">"just_your_ordinary_fasta_file.fasta"</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Dominik Rafacz, Michal Burdukiewicz, Mateusz Bakala.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
